Allow FileSystemProvider to stat a file as readonly (#73122)

gjsjohnmurray opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests.
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `master` branch.
* Include a description of the proposed changes and how to test them.
-->

This draft PR addresses #73122

It adds an optional `readonly` boolean property to the FileStat interface, thus enabling a FileSystemProvider to control whether a file is editable or read-only.

One way to test:
1. Clone https://github.com/gjsjohnmurray/vscode-extension-samples/tree/test-vscode-73122-fsp-stat-readonly
2. Open fsprovider-sample folder.
3. Add this PR's vscode.proposed.d.ts file to that folder.
4. Build and run.
5. Run the commands to init the MemFS workspace and add the MemFS files.
6. Open any of the .txt files to see them read-only.
7. Open any of the other files to see that they are editable as usual.

Submitted as a draft PR seeking comment and guidance from @jrieken about taking this through the API proposal precedure.

jrieken commented on src/vs/vscode.proposed.d.ts:
```diff
@@ -2163,6 +2163,47 @@ declare module 'vscode' {
+	 * The `FileStat`-type represents metadata about a file
+	 */
+	export interface FileStat {
+		/**
```
Repeating all existing members of the interface shouldn't be required as TS merges them based on their name.

jrieken commented on src/vs/workbench/api/common/extHostFileSystem.ts:
```diff
@@ -115,6 +115,7 @@ export class ExtHostFileSystem implements ExtHostFileSystemShape {
 	private readonly _fsProvider = new Map<number, vscode.FileSystemProvider>();
 	private readonly _registeredSchemes = new Set<string>();
 	private readonly _watches = new Map<number, IDisposable>();
+	private readonly _enableProposedApi = new Map<number, boolean>();
```
not a super-fan of this extra data structure (which doesn't seem to be clean up btw). could be consider debt tho and cleaned up later

jrieken commented:
Adding @mjbvz for custom editors

jrieken commented:
> Submitted as a draft PR seeking comment and guidance from @jrieken about taking this through the API proposal precedure.

The checks are in place and things are defined proposed.d.ts - so all good from that side. All API proposals are discussed during a weekly call and I can represent this issue. So far, the FS provider API has been posix-like and this isn't strictly like it anymore, tho personally I don't mind it.

bpasero commented on src/vs/platform/files/common/files.ts:
```diff
@@ -742,13 +747,22 @@ interface IBaseStat {
+	 * The value may or may not be resolved as
+	 * it is optional.
+	 */
+	readonly?: boolean;
```
`readonly` should only be optional in `IStat` but given it is included in each `IStat` it can be required on the level of `IBaseStat`, no?

gjsjohnmurray replied:
Changes pushed. Please verify I understood you correctly.

gjsjohnmurray replied:
@bpasero I am now unsure about the wisdom of making `readonly` required in IBaseStat, IBaseStatWithMetaData and IFileStatWithMetaData. It mean that if a FSP doesn't define the new `readonly` property of FileStat I default it true or false depending on whether the FSP registered as a readonly one or not. But if the FSP subsequently unregisters and reregisters with the opposite readonly mode (now supported, thanks to your #110854), already-resolved files in the FSP report the previous default value for readonly.

I enhanced my fsprovider-sample branch to allow exploration of this.

bpasero commented on src/vs/workbench/services/textfile/common/textFileEditorModel.ts:
```diff
@@ -31,6 +31,7 @@ interface IBackupMetaData {
 	size: number;
 	etag: string;
 	orphaned: boolean;
+	readonly?: boolean;
```
I would not add `readonly` to backups as I see no benefit doing so.

gjsjohnmurray replied:
Removed in latest push.

bpasero commented on src/vs/workbench/services/textfile/common/textFileEditorModel.ts:
```diff
@@ -1028,6 +1031,9 @@ export class TextFileEditorModel extends BaseTextEditorModel implements ITextFil
 	}
 
 	isReadonly(): boolean {
+		if (this.lastResolvedFileStat?.readonly) {
```
This is wrong, no? I think if we have a `boolean` `readonly` property we have to respect that independent from the file system? Because someone might contribute a readonly file system where selected files are not readonly?

gjsjohnmurray replied:
Fair point. But if someone is already contributing a readonly filesystem whose `stat` doesn't return any value for the file-level `readonly` property (because it didn't exist), we must default it to `true` for a readonly FS and `false` otherwise, right?

bpasero replied:
Yeah, only take it unless `undefined`

gjsjohnmurray replied:
Please review the change I made in latest push.

bpasero reviewed (changes requested):
Left some feedback in the code. On top of that:
* isn't [`mainThreadFileSystem#$stat`](https://github.com/microsoft/vscode/blob/9f9ba405e4d1855161a22d86da8b705753147089/src/vs/workbench/api/browser/mainThreadFileSystem.ts#L63-L70) missing to set the `readonly` flag?
* this should get a test to verify the text editor model reports readonly for readonly file stat

gjsjohnmurray commented:
> 
> 
>     * isn't `mainThreadFileSystem#$stat` missing to set the `readonly` flag?

Added in latest push.

> 
>     * this should get a test to verify the text editor model reports readonly for readonly file stat

I'd appreciate some guidance on doing this.


I have updated my branch of fsprovider-sample so the readonly state of a file can be toggled while open. If I close the file tab and reopen it the readonly state of the editor changes, but I think this needs to happen dynamically. @bpasero you recently did something similar in https://github.com/microsoft/vscode/issues/110854 to handle a FSP's readonly capability changing on the fly.

gjsjohnmurray commented:
I am back at my desk after the holidays, so I've removed draft status from this PR.

bpasero commented:
@gjsjohnmurray sorry for the time this took. I tried to bring your branch up to date with `main` but I do not have permissions it seems to push. Anyway, maybe you can force your branch to be on this commit and then we can proceed here: https://github.com/microsoft/vscode/commit/cf122a220813a5e46d0b14e36eb335f99ba88345

**Remaining tasks:**
* test coverage (aka ensure the readonly change event fires when a underlying provider changes readonly state for a file) [1]
* make sure the editor control updates properly based on readonly changes, similar to how we currently do it when the provider capabilities change ([2]). A provider that after some time decides to change the `readonly` state for a file should reflect in the editor turning readonly or not

[1] as for best ways how to test this, I would suggest to look at the `src/vs/workbench/services/workingCopy/test/browser/fileWorkingCopy.test.ts`. file working copies will eventually be the default for text file model too, so we do not have to add tests for both. I would think that the coverage should be:
* the `onDidChangeReadonly` event firing properly
* the `isReadonly` respecting the last resolved file stat or fallback to the system provider readonlyness

[2] https://github.com/microsoft/vscode/blob/f4ab7e822f98611f54f3972b451836750e74f65e/src/vs/workbench/contrib/files/browser/editors/textFileEditor.ts#L83-L89

gjsjohnmurray commented:
> I tried to bring your branch up to date with `main` but I do not have permissions it seems to push. Anyway, maybe you can force your branch to be on this commit and then we can proceed here: [cf122a2](https://github.com/microsoft/vscode/commit/cf122a220813a5e46d0b14e36eb335f99ba88345)

@bpasero I'm not familiar with this procedure. I have now set the PR's checkbox `Allow edits and access to secrets by maintainers`, so maybe you can do it yourself now. Or else please tell me the git command to issue or the option to pick from the SCM ... menu in VS Code.

bpasero commented:
@gjsjohnmurray done

bpasero commented:
@gjsjohnmurray are you planning to work on this further? Otherwise I can look into the finishing touches.

From a API discussion we had today, we picked one of the proposals from https://github.com/microsoft/vscode/issues/73122#issuecomment-838025028. Basically it would be this:

```ts
export enum FileStatPermissions {
	/**
	 * The file is readonly.
	 */
	Readonly = 1
}

/**
 * The `FileStat`-type represents metadata about a file
 */
export interface FileStat {

	/**
	 * The permissions of the file, e.g. whether the file is readonly.
	 *
	 * *Note:* This value might be a bitmask, e.g. `FileStatPermissions.Readonly | FileStatPermissions.Other`.
	 */
	permissions?: FileStatPermissions;
}
```

gjsjohnmurray commented:
@bpasero I'll happily leave it in your safe hands. Thanks!

bpasero commented:
Thanks 🥂
